### ISSUE TITLE
draft requirements for new NXvalidate

### ISSUE THREAD

**Issue by prjemian** [nexusformat/communications#17]
Mark started a draft stating requirements for a new version of NXvalidate library and application.

On Thu, Oct 8, 2015 at 10:09 AM, Koennecke Mark (PSI)
<mark.koennecke@psi.ch <mailto:mark.koennecke@psi.ch>> wrote:

    Hi,

    I have written a little white paper about how I would reimplement
    NXvalidate. It details what needs to be
    done and a little how. There are a number of TODO items in the text.
    Please have a look. May be we
    can sort some of this out during one of our next Telcos. Then I
    might start hacking away at this.
    Otherwise this may be Code Camp material. The txt is the markdown
    original if someone feels like
    editing the text.

    Best Regards,

          Mark Koennecke

**Comment by prjemian** [comment]
see my changes to Mark's initial draft:
https://github.com/nexusformat/communications/commit/cab0e2e5cf5ec3cd34223c4a5640b88deb70f5ce

**Comment by mkoennecke** [comment]
Hi,

it might be better to put the comments in here rather then to write them into the paper.
A brief summary as of now:

* There are concerns about efficiency, thus there seems to be an agreement to to implement this 
  in C or C++. Where I have a preference for C. After decades of pain I do not trust C++ compilers 
 anymore... Moreover C seems to far easier to wrap for other programming languages then C++
* I support Petes argument that you do not need to validate each file. Typically I would use 
  NXValidate after making changes to file writing code in the DAQ in order to verify that I did 
  the right thing.
* Pete argues to use libxml2 rather them MXML. We could do. The reason why I used MXML for 
  the NAPI-XML had to do with efficiency in loading arrays of numbers. This reason does not 
  apply for NXvalidate.

I myself had second thoughts about the API and changed it to a C with objects style of API. 
It should now also be safe enough to have multiple instances of NXvalidate running, another 
help with efficiency. 
 
Mark Koennecke

**Comment by eugenwintersberger** [comment]
I'd like to make some remarks on the current state of Mark's white paper with a focus on technical issues. 

### Which platforms and architectures 

I think an important question which is not answered by the white paper currently is which platforms (less important, which architectures) we want to support. I guess the following platforms are mandatory 
* Linux 
* OSX
* Windows
We definitely should answer this question before discussion all other technical issues (libraries, languages). 

### Usage of NAPI 

According to the white paper, Mark prefers NAPI over the native HDF5 or any other API. Though that makes sense from his point of view, deciding to use NAPI for the development of a new tool would contradict our original statement that NAPI is in maintenance mode and is not recommended for being used in new software projects. 

### C vs. C++

I agree with Mark that the functionality should be implemented in a library. As a matter of fact Mark is right when he favors C as the basic language for implementing the library (also the white paper also states that C++ would provide a far more cleaner interface). From my personal experience C++ is indeed a better choice as it is far easier to get robust code in C++ rather than in C. Concerning bindings to other languages C++ is not as bad as it seems 
* C++ to Python (boost::python or shiboken)
* C++ to Java (JNI - there is also some other library but I need to ask our Java gurus for this)
* C++ to C - not a problem (in fact I would rather go this direction than the other way around).

Are there any other languages we want to provided bindings for?

### Usage of libxml2 

Would be fine if we stay in stay in the Linux/Unix universe. If we do not need a validating parser boost::property_tree would be a reasonable choice. When going for C++ Boost most probably will be a dependency anyhow and we know that it works on Windows. The currently available binaries for libxml2 on Windows are from 2011 and 2009 which is a clear indicator for me that the there is not too much active development. 

That's it for the moment. I will come back when I have spent more time thinking on the other issues mark has mentioned in his white paper.

**Comment by rayosborn** [comment]
On Oct 11, 2015, at 5:31 AM, Eugen Wintersberger <notifications@github.com<mailto:notifications@github.com>> wrote:
Usage of NAPI

According to the white paper, Mark prefers NAPI over the native HDF5 or any other API. Though that makes sense from his point of view, deciding to use NAPI for the development of a new tool would contradict our original statement that NAPI is in maintenance mode and is not recommended for being used in new software projects.

This is a point I have made in the past, but I believe it more strongly now with recent experience  - that the NeXus format needs to serve more than the software developers at large-scale facilities, which seems to be the basis for many recent decisions. We want application developers to write NeXus files as an export option and we want scientists, who write the occasional program, to save their results in NeXus. I don’t believe that this will be possible if we expect them to use the native HDF5 library to follow the increasingly arcane rules that we have recently agreed on. If anything, we need to increase the functionality of NAPI to make it simpler to create, for example, a standard-conforming NXdata group.

If you ask someone who doesn’t want to spend their life reading both the NeXus spec and the HDF5 documentation to write a NeXus file, they will get it wrong. I have read plenty of examples using the Python API and spent a lot of time writing scripts to correct them. It is my belief that, if Mark and others are able to write a version of NAPI that only applies to HDF5, it should not be hard to maintain. Most of the members of the committee have probably already written their own versions of the NAPI in native HDF5 - we just need to agree on the best way of pooling those versions.

Ray
--
Ray Osborn, Senior Scientist
Materials Science Division
Argonne National Laboratory
Argonne, IL 60439, USA
Phone: +1 (630) 252-9011
Email: ROsborn@anl.gov<mailto:ROsborn@anl.gov>

**Comment by prjemian** [comment]
The last time the NIAC members were polled on their use of the NAPI, 
less than 50% said they used it to read or write data.

**Comment by eugenwintersberger** [comment]
I agree with Ray so far as using the native HDF5 API is not a good option, in particular not for scientists which I typically consider as casual programmers. However, in its current state, NAPI is simply outdated (opening and closing objects for a single read/write access is not really what you want). The only way to repair this would be a total rewrite. This is why I started my own C++ API when I started my work at DESY. 

But I think I have already once suggested to provide a list of links to currently developed NeXus APIs on our website ;)

**Comment by rayosborn** [comment]
Pete, if only the NIAC members were polled, then that doesn't affect anything I wrote. I agree with Eugen that there could be a case for not having a state-based API. We have discussed in the past having a higher-level API, but I am not up-to-date with where that discussion ended. Is that on the Code Camp agenda?

**Comment by zjttoefs** [comment]
I'm not sure how this mix of editing the document and the comments in this ticket work. The white paper contradicts itself almost on whether validating depends_on is required (it should be). 

Before we enter the technical details I would suggest to closely look at the requirements. 
How much higher lever code on top of (or along side of) NXDL will we need? 
How does that align with the experimental "features" project we have? 
How do we deal with multiple versions of a definition?

**Comment by mkoennecke** [comment]
Hi,

there are a number of issues which have come up here:

* NAPI versus HDF5 API. Now, if the requirement is that we validate also HDF-5 files which might 
  not be NeXus compliant (missing NeXus attributes like file attributes or NX_class etc), then this 
  requirement forces the use of the HDF-5 API. Else we can use NAPI. It is that simple. Decide on 
  this requirement, choice made.
* NAPI versus HDF-5 again: What we use to implement NXvalidate is completely, I repeat completely, 
  decoupled from the future of the NAPI. NXvalidate will be coded by us, the experts in both. 
* C versus C++: This is a religious war. We can fence it out with drinking straw at the code camp :-)
  And then go away and code it in GO..... I am still for C; and it includes the C++ camp as C++ contains 
 C....
* The question of platform was raised: Linux for sure, OSX is a byproduct of Linux but I think, 
  unfortunately, we have to do Windows as well.  
* Choice ox XML library: Either libxml2 or MXML are good for me. And if we do windows, we have to 
  do an installer anyway. And that installer can also take care of installing the XML library of our choice.
* We have the same issue about choosing a XML library with C++ too: the standard library, even 
  boost, does not provide anything. 
* I have been misunderstood: I am more or less asking in the document if depends_on chains are 
  to be validated. The answers appears to be: yes, we want that
* Dealing with multiple versions of NXDL:  the default operation it to use the NXDL encoded in the 
  definitions field of NXentry or NXsubentry. If you want something else you have to give the validator 
  the NXDL file as an argument
* Now features: I always thought that as another project in python, independent of NXvalidate. Having 
  said that, any feature which can be expressed in NXDL can also be validated by NXvalidate. If we 
  go beyond that, for example feature logic, we would need a description for that. That would 
  complicate NXvalidate beyond measure. We want that? A possible course of action would be to call 
  an external program in order to verify logic not implemented in NXvalidate.
* The units validation issue was solved by Pete.

Thats it for now,

    Mark Koennecke

**Comment by zjttoefs** [comment]
Just quickly:

Versions: How does NXvalidate retrieve old versions? Does it clone the repository locally?
Features: The main point was to allow better validation. Does not have to be integrated. But then having any high level logic in validate might not be required.

Tobias

**Comment by rayosborn** [comment]
On Oct 12, 2015, at 9:32 AM, mkoennecke <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi,

there are a number of issues which have come up here:

  *   NAPI versus HDF5 API. Now, if the requirement is that we validate also HDF-5 files which might not be NeXus compliant (missing NeXus attributes like file attributes or NX_class etc), then this requirement forces the use of the HDF-5 API. Else we can use NAPI. It is that simple. Decide on this requirement, choice made.

Hi Mark,
I’m not 100% sure I understand what you are saying here. Are you saying that we should use the NAPI in all cases that don’t require a validation step?

The Python API will read in any HDF5 file, compliant or not. If there is no NX_class attribute, the group is just given a generic NXgroup class. So validation can be built on top of the API without needing to use h5py directly. I guess in principle, the NAPI could operate the same way.

Ray
--
Ray Osborn, Senior Scientist
Materials Science Division
Argonne National Laboratory
Argonne, IL 60439, USA
Phone: +1 (630) 252-9011
Email: ROsborn@anl.gov<mailto:ROsborn@anl.gov>

**Comment by mkoennecke** [comment]
Ray,

I am saying: what we use for reading and writing NeXus in normal cases is completely  independent of 
the implementation of NXvalidate. Now, people now write NeXus files with the HDF-5 API alone which 
might not be NeXus level 1 compliant. For instance, not even have the NX_class attribute on 
groups. If we want to validate this, the NAPI stands in the way and we better use the HDF-5 API for 
NXvalidate. 

Mark K

**Comment by prjemian** [comment]
For me, the question is whether or not the NAPI can report that any given
data file is or is not a valid NeXus file, let alone valid by one or more
specific application definitions.

The previous version of nxvalidate failed on this point.  Because it relied
on nxconvert, it could only report on files that passed that step.  It
errored out on any file that failed nxconvert, and with no reported message
about why it failed.

This type of error MUST NOT HAPPEN in a file validation program.

**Comment by rayosborn** [comment]
Mark, 
I thought that was what you meant. I agree with that.

Pete,
As Mark says, we have to be careful to keep the use of the NAPI and NXvalidate as two separate discussions. So we don't necessarily need the NAPI to do validation, although it will probably throw an error if it is doesn't even define group classes.

**Comment by prjemian** [comment]
Ray

I fully agree with you.  In validation code, we will never write a data
file.  Whatever code we use must be robust for any type of pathology in the
file.  The old validation code was not robust for the most basic error.

The new validation code must consider the case where the person presenting
the data file is not sure of the NeXus rules and misses fundamental things,
such as the NX_class attribute on groups.

**Comment by prjemian** [comment]
libxml2: http://xmlsoft.org/

**Comment by prjemian** [comment]
Windows libxml2:  https://www.zlatkovic.com/libxml.en.html

**Comment by prjemian** [comment]
mxml:  https://en.wikipedia.org/wiki/MXML

**Comment by prjemian** [comment]
Microsoft Core XML Services: http://www.microsoft.com/en-us/download/details.aspx?id=3988

**Comment by mkoennecke** [comment]
Hi,

I have pushed a first version of c-nxvalidate today. It is incomplete. Many tests and features are missing. But the framework is there. It is in a state where others could join in and implement more 
tests. Just search for the TODO scattered all over the code. 

I choose to use libxml2 as it is more powerful and on the same level of support as mxml.

I am tempted to drop the warning levels from the validator: If someone wants to filter the validators 
output she should implement a suitable logging function. 

I am still looking for an elegant way to validate attributes. The problem being that attributes can come in two forms: as proper XML attributes to field or groups in which case they have data. Or as 
attribute nodes which may or may not have data. But do not worry, a hideous way to solve this with much typing is always possible.    

I got disenchanted with the maturity of cmake. I have at least three bug reports against cmake now...

Then there is the matter of automatic testing. We should have a test suite. Question is if we keep that at the smoke test level or if we do better. Which raises the issue of which testing tool to use. 
I am currently considering https://github.com/philsquared/Catch because it supports the BDD style of testing I like these days and comes as a single header file which we could distribute. 
Or do we write do a python wrapper right away and use Behave? The approach has the advantage that it is easier to produce test data and xdl files on the fly rather then carrying them around.
